DataFormats/PatCandidates/interface/PATObject.h

```diff
@@ -717,123 +726,167 @@ namespace pat {
-                overlapItems_[match - overlapLabels_.begin()] = overlaps;
-            }
-        }
+      
+        auto match = std::lower_bound(overlapLabels_.begin(), overlapLabels_.end(), label);
+        const auto dist = std::distance(overlapLabels_.begin(),match);
+        if ( match == overlapLabels_.end() || *match != label ) {
+          overlapLabels_.insert(match,label);
+          overlapItems_.insert(overlapItems_.begin()+dist,overlaps);
+        } else {
+          overlapItems_[dist] = overlaps;
+        }        
   }
 
   template <class ObjectType>
   const pat::UserData * PATObject<ObjectType>::userDataObject_( const std::string & key ) const
   {
-    std::vector<std::string>::const_iterator it = std::find(userDataLabels_.begin(), userDataLabels_.end(), key);
-    if (it != userDataLabels_.end()) {
-        return & userDataObjects_[it - userDataLabels_.begin()];
+    auto it = std::lower_bound(userDataLabels_.cbegin(),userDataLabels_.cend(),key);
+    if ( it != userDataLabels_.cend() && *it == key) {
+      return &userDataObjects_[std::distance(userDataLabels_.cbegin(),it)];
     }
-    return 0;
+    return nullptr;
   }
 
   template <class ObjectType>
   float PATObject<ObjectType>::userFloat( const std::string &key ) const
   {
-    std::vector<std::string>::const_iterator it = std::find(userFloatLabels_.begin(), userFloatLabels_.end(), key);
-    if (it != userFloatLabels_.end()) {
-        return userFloats_[it - userFloatLabels_.begin()];
+    auto it = std::lower_bound(userFloatLabels_.cbegin(),userFloatLabels_.cend(),key);
+    if( it != userFloatLabels_.cend() && *it == key ) {
+      return userFloats_[std::distance(userFloatLabels_.cbegin(),it)];
     }
-    return 0.0;
+    // cuts based on non-existing keys should always fail, so use NaN, not zero
+    cms::Exception ex("UnknownUserFloat");
+    ex << "Requested UserFloat " << key << " is not available! Possible UserFloats are: " << std::endl;
+    for( const auto& name : userFloatLabels_ ) {
+      ex << name << ' ';
+    }
+    throw ex;
+    return std::numeric_limits<float>::quiet_NaN(); 
   }
 
   template <class ObjectType>
   void PATObject<ObjectType>::addUserFloat( const std::string & label,
-					    float data )
+					    float data,
+                                            const bool overwrite )
   {
-    userFloatLabels_.push_back(label);
-    userFloats_.push_back( data );
+    auto it = std::lower_bound(userFloatLabels_.begin(),userFloatLabels_.end(),label);
+    const auto dist = std::distance(userFloatLabels_.begin(),it);
+    if( it == userFloatLabels_.end() || *it != label ) {      
+      userFloatLabels_.insert(it,label);
+      userFloats_.insert(userFloats_.begin()+dist,data);
+    } else if( overwrite && *it == label  ) {
+      userFloats_[dist] = data;
+    } else {
+      edm::LogWarning("addUserFloat") 
+        << "Attempting to add userFloat: " 
+        << label << " = " << data << " failed, value exists already!";
+    }
```

Comment:
It was fairly easy to access the extra pairing if anybody was inheriting from the PAT objects (the user float data is protected, not private), which may be done in some analysis frameworks.

That said, the old behaviour was bugged because it was written in 5 minutes without much thought. Now that you're rewriting it, I think it makes more sense to spend another 10 minutes reviewing it and implementing it in a more complete form, rather than having a version which, if run with default parameters, behaves possibly even worse than the old one.